move/committee/sources/committee.move

```diff
@@ -0,0 +1,908 @@
+
+    transfer::share_object(Committee {
+        id: object::new(ctx),
+        threshold,
+        members: new_members,
+        state: State::Init { candidate_data: vec_map::empty() },
+        old_committee_id: option::some(old_committee.id()),
+    });
+}
+
+/// Register a candidate with ecies pk, signing pk and URL. Append it to
+/// candidate data map.
+public fun register(
+    committee: &mut Committee,
+    enc_pk: vector<u8>,
+    signing_pk: vector<u8>,
+    url: String,
+    ctx: &mut TxContext,
+) {
+    // TODO: add checks for enc_pk, signing_pk to be valid element, maybe PoP
+    assert!(committee.members.contains(&ctx.sender()), ENotMember);
+    let sender = ctx.sender();
+    match (&mut committee.state) {
+        State::Init { candidate_data } => {
+            assert!(!candidate_data.contains(&sender), EAlreadyRegistered);
+            candidate_data.insert(sender, CandidateData { enc_pk, signing_pk, url });
+        },
+        _ => abort EInvalidState,
+    }
+}
+
+/// Propose a committee with a list partial pks (in the order of committee's members list)
+/// and master pk after DKG. Add the caller to approvals list.
+/// If already in PostDKG state, check the submitted partial_pks and pk are consistent with
+/// the onchain state, then add the caller to approvals list.
+public fun propose(
+    committee: &mut Committee,
+    partial_pks: vector<vector<u8>>,
+    pk: vector<u8>,
+    ctx: &mut TxContext,
+) {
+    assert!(committee.members.contains(&ctx.sender()), ENotMember);
+    assert!(partial_pks.length() == committee.members.length(), EInvalidProposal);
+
+    match (&mut committee.state) {
+        State::Init { candidate_data } => {
+            assert!(candidate_data.length() == committee.members.length(), ENotRegistered);
+            let mut i = 0;
+            while (i < committee.members.length()) {
+                assert!(candidate_data.contains(&committee.members[i]), ENotRegistered);
+                i = i + 1;
+            };
+
+            committee.state =
+                State::PostDKG {
+                    candidate_data: *candidate_data,
+                    approvals: vec_set::singleton(ctx.sender()),
+                    partial_pks,
+                    pk,
+                };
+        },
+        State::PostDKG {
+            approvals,
+            candidate_data: _,
+            partial_pks: existing_partial_pks,
+            pk: existing_pk,
+        } => {
+            // Check that submitted partial_pks and pk are consistent
+            assert!(partial_pks == *existing_partial_pks, EInvalidProposal);
+            assert!(pk == *existing_pk, EInvalidProposal);
+
+            // Insert approval and make sure if approval was not inserted before
+            assert!(!approvals.contains(&ctx.sender()), EAlreadyProposed);
+            approvals.insert(ctx.sender());
+        },
+        _ => abort EInvalidState,
```

Comment:
why not silent success for `Finalized`?

Reply:
since now finalize is part of propose, the only valid state for propose is in init or postdkg, hence keeping this invalid state error.